common/lib/util/defaults.js

```diff
@@ -116,16 +127,22 @@ Defaults.normaliseOptions = function(options) {
 	if(!('queueMessages' in options))
 		options.queueMessages = true;
 
-	var production = false;
 	if(options.restHost) {
 		options.realtimeHost = options.realtimeHost || options.restHost;
+		options.fallbackHosts = options.fallbackHostsUseDefault && !options.port && !options.tlsPort ? Defaults.FALLBACK_HOSTS : options.fallbackHosts;
 	} else {
 		var environment = (options.environment && String(options.environment).toLowerCase()) || Defaults.ENVIRONMENT;
-		production = !environment || (environment === 'production');
+		var production = !environment || (environment === 'production');
 		options.restHost = production ? Defaults.REST_HOST : environment + '-' + Defaults.REST_HOST;
 		options.realtimeHost = production ? Defaults.REALTIME_HOST : environment + '-' + Defaults.REALTIME_HOST;
+		if(!options.fallbackHosts && !options.port && !options.tlsPort) {
+			if(production || options.fallbackHostsUseDefault) {
```

Comment:
Just an ambiguous question ... user specifies environment (sandbox) with fallbackHostsUseDefault as true. After disconnection with the primary host (sandbox-ably.realtime.com).. subsequent redirects will be with the wrong hosts. I think although fallbackHostsUseDefault is deprecated, it's wrong for the library allowing to do that.

Reply:
I mean, nothing serious. But, it's still something the library can manage

Reply:
We can close this conversation if it's already well thought : )

Reply:
> After disconnection with the primary host (sandbox-ably.realtime.com).. subsequent redirects will be with the wrong hosts.

Yes, if you tell the library to use the fallback hosts for the `production` cluster when your primary endpoint is a different cluster, then the library will be broken. The solution is to not do that - `fallbackHostsUseDefault` is for when using a primary endpoint that's just an alias for the production cluster. The library has no way of knowing when that is the case based on static analysis of the options only, we don't want the lib to have to make http requests to decide whether the options are valid. If you're giving fallback options the library is entitled to assume you know what you're doing. (Anyway, we are deprecating that option soon in favour of creating dns entries for the full set of autogenerated fallbacks)

Reply:
As Simon said, the library has no way to infer whether these options are compatible or not, so it is up to the user to do the right thing, but I have at least added a warning to make it more apparent since there is no longer a need to set both anyway: https://github.com/ably/ably-js/pull/689